### DISCUSSION TITLE
BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES

### DISCUSSION THREAD

**Discussion by zsh151** [neams-th-coe/cardinal#483] [Q&A] [answered]
Hi all,
I was trying to run a thinly-wrapped NekRS simulation without any physics coupling following Tutorial 1 (tutorials/standalone).
There are some errors that I can't understand.
```
===================================================================================
=   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
=   PID 8990 RUNNING AT ubuntu
=   EXIT CODE: 9
=   CLEANING UP REMAINING PROCESSES
=   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
===================================================================================
YOUR APPLICATION TERMINATED WITH THE EXIT STRING: Killed (signal 9)
This typically refers to a problem with your application.
Please see the FAQ page for debugging suggestions
```

The following text is the entire output. 
[output.txt](https://github.com/neams-th-coe/cardinal/files/9706596/output.txt)

Thanks.

**Comment by aprilnovak** [comment, marked as answer]
Hi @zsh151 - are you running an existing case, or are you starting from scratch? Whenever you see a "BAD TERMINATION..." that means that the simulation crashed, and the actual error is going to appear just before this message. Unfortunately in this case there is no error message, but I can see that the simulation is failing in `calling udf_setup ...`. 

Can you paste your `UDF_Setup` function here? This function is in `<case>.udf`.

**Reply by zsh151** [reply]
I was running the tutorials case (cardinal/tutorials/standalone). That is the turbPipe case too.

The following is the contains in the turbPipe.udf.
```
#include <math.h>
#include "udf.hpp"
#include "plugins/velRecycling.hpp"
#include "plugins/avg.hpp"

#define ZLENGTH  20.0  /* diameters  */
#define NSLABS 60
#define NELSLAB 108

void UDF_LoadKernels(nrs_t *nrs)
{
  velRecycling::buildKernel(nrs);
  avg::buildKernel(nrs);
}

void UDF_Setup(nrs_t *nrs)
{
  // setup scratch space
  platform_t* platform = platform_t::getInstance();
  nrs->o_usrwrk = platform->device.malloc(nrs->NVfields*nrs->fieldOffset,  sizeof(dfloat));

  // recycling
  const dfloat wbar  = 1.0;
  const int bID      = 1;
  dfloat zRecycLayer = 0.25*ZLENGTH;
  const hlong offset = NELSLAB * round(NSLABS * zRecycLayer/ZLENGTH);
  velRecycling::setup(nrs, nrs->o_usrwrk, offset, bID, wbar);

  avg::setup(nrs);
}

void UDF_ExecuteStep(nrs_t *nrs, dfloat time, int tstep)
{
  mesh_t *mesh = nrs->meshV;

  velRecycling::copy();
  avg::run(time);

  if (nrs->isOutputStep) {
    occa::memory o_UZ = nrs->o_U + 2*nrs->fieldOffset * sizeof(dfloat);
    const dfloat ubar = platform->linAlg->innerProd(mesh->Nlocal, o_UZ, mesh->o_LMM, platform->comm.mpiComm)/mesh->volume;
    if (platform->comm.mpiRank == 0) printf(" uBulk: %g\n", ubar);

    avg::outfld();
  }
}
```

**Reply by aprilnovak** [reply]
Hmm, that should run fine - we test the tutorials as part of our test suite. Can you try just running the NekRS executable so that we can narrow down if this is a Nek issue vs. a Cardinal issue?

```
cardinal/install/bin/nrsmpi turbPipe 4
```

**Reply by zsh151** [reply]
Well, it seems the same problem appears when I run
```
nrsmpi turbPipe 4
```
[output.txt](https://github.com/neams-th-coe/cardinal/files/9712195/output.txt)

**Reply by aprilnovak** [reply]
Ok can you try one last thing just to be sure - sometimes an old `.cache` messes with things:

```
cd tutorials/standalone
rm -rf .cache/
nrsmpi turbPipe 4
```

If that doesn't work, then this is almost certainly an issue with NekRS itself, and not Cardinal. Can you run any other Cardinal cases? Perhaps it is just something about the turbPipe example (which we should report to the NekRS team if it persistently doesn't work on your system).

```
cd test/tests/cht/sfr_pincell
cardinal-opt -i nek_master.i
```

**Reply by zsh151** [reply]
Thanks.
These scripts cannot work either.
But the sfr_pincell case could run smoothly.

**Reply by aprilnovak** [reply]
Ok thanks for reporting this! I will look into this and hopefully have a fix soon. My guess is that there's some function being called from the `turbPipe` case (which we just exactly copied from an existing NekRS example), that doesn't work nicely on your system. You should be fine to move forward with building cases based on other tests and tutorials though.

**Reply by zsh151** [reply]
Okay, thanks for your answering.
